Restore deterministic r.now

AtnNn opened the pull request:
Fixes #6125

Replaces the pseudo-boolean `deterministic_t` with a bitmask.

srh commented on src/rdb_protocol/order_util.cc:
```diff
@@ -33,7 +33,7 @@ build_comparison(const term_t* target,
     counted_t<const func_t> comparison_function = arg->as_func(GET_FIELD_SHORTCUT);
     rcheck_target(target,
-                  comparison_function->is_deterministic() != deterministic_t::no,
+                  (comparison_function->is_deterministic() &~ DET_SINGLE_SERVER) == DET_DETERMINISTIC,
```
Sorry, I don't know of any &~ operator in C++.

srh replied:
So this was saying the value is either single-server or deterministic.

But now in a convoluted way, it's saying that the value is either DET_SINGLE_SERVER or DET_DETERMINISTIC, but not DET_CONSTANT_NOW.

AtnNn replied:
And not DET_NONDET either.

srh commented on src/rdb_protocol/term.hpp:
```diff
@@ -21,7 +21,7 @@ class table_t;
 class var_captures_t;
 class compile_env_t;
-enum class deterministic_t;
+typedef int deterministic_t;
```
Instead this could be wrapped in a struct?

AtnNn replied:
Done.

srh commented on src/rdb_protocol/op.cc:
```diff
@@ -356,5 +352,21 @@ bool bounded_op_term_t::open_bool(
     }
 }
 
+std::string deterministic_string(deterministic_t det) {
```
Nothing uses this function `deterministic_string`.

AtnNn replied:
I renamed it to `debug_print`

srh commented on src/rdb_protocol/op.hpp:
```diff
@@ -18,22 +18,24 @@ namespace ql {
 
-// Note that worst_determinism depends on the order of these; if you want to add
-// a new value, make sure worst_determinism is properly defined for it.
-enum class deterministic_t {
```
Is it now possible for `DET_NONDET | DET_SINGLE_SERVER` to be a value of deterministic_t?  Is it possible for `DET_NONDET` to be a value?

AtnNn replied:
Yes

srh commented:
I think the main problem afaict is that deterministic_t has multiple values that mean the same thing.  The encapsulation of deterministic_t doesn't really do anything because it's still bitmasking stuff.

AtnNn commented:
@srh `deterministic_t` is a set of possibly non-deterministic operation types. I use a bitset to represent them because `std::set` would be overkill.

It used to be an enum with a `worse_determinism` ordering, but that ordering won't be total anymore because of the introduction of non-deterministic `r.now`.

Given `det` the set of operation types present in a query and `mask` the set of deterministic operations in the current environment, the set `det & ~mask` is empty if the query is deterministic.

Because `mask` will never contain `NONDET`, some values of `det` will mean the same thing, but I don't think that's a problem.

srh commented:
I think that doing it this way is really confusing.  It's hard to keep track of what it means when these bitmask operations are done, and it's hard to understand what the meaning of an assert_deterministic(MASK, ...) call is supposed to mean.  You have to negate and bitwise and stuff in your head.

Could is_deterministic take a boolean indicating whether r.now() terms are to be deemed deterministic, and then produce a {no,single_server,always} result?

srh commented on src/rdb_protocol/terms/seq.cc:
```diff
@@ -461,18 +461,10 @@ struct rcheck_transform_visitor_t : public bt_rcheckable_t,
-                          "Cannot call `changes` after a non-deterministic function.");
-
-            default: unreachable();
+        if((f.compile_wire_func()->is_deterministic() & ~deterministic_t::SINGLE_SERVER) != deterministic_t::DETERMINISTIC) {
```
Why is r.now() non-constant here?  (I have no idea whether it should be.)

srh commented on src/rdb_protocol/order_util.cc:
```diff
@@ -33,7 +33,7 @@ build_comparison(const term_t* target,
     counted_t<const func_t> comparison_function = arg->as_func(GET_FIELD_SHORTCUT);
     rcheck_target(target,
-                  comparison_function->is_deterministic() != deterministic_t::no,
+                  (comparison_function->is_deterministic() & ~deterministic_t::SINGLE_SERVER) == deterministic_t::DETERMINISTIC,
```
Why does this not permit r.now() to be constant?  (I have no idea whether it should.)

srh commented:
I made a branch at https://github.com/srh/rethinkdb/commits/sam/deterministic_now which adds a few commits, that gives deterministic_t a tighter interface (and fixes a compile bug, and makes the names more like what they were originally).  The most important difference is that it removes operator== and operator!= so that functionally identical deterministic_t values are no longer distinguishable.

It doesn't address my last two questions, about the correctness of changes in order_util.cc or seq.cc.

thelinuxlich commented:
#3583

srh commented:
That's a ReQL proposal that this change partially implements.  `r.now()` becomes reevaluated for each change in .changes().  Note that as implemented it's *not* the timestamp at which the change happens -- it's some other timestamp that gets computed later.  That's pretty weird.

The other change that this introduces is that r.now() can't be used *before* .changes() is called.  If you want changes that refer to anything referring to a future time, or a past time, or anything over 3 weeks old, now you can't do that using r.now().

srh commented:
Another thing that's broken is queries that sort using r.now().

For example, if you want to order rows by the absolute distance of the time from some point in history:
`r.table('foo').orderBy(r.row('time').sub(r.now()).do(function (x) { return r.branch(x.gt(0), x, r.expr(0).sub(x)); }))`

thelinuxlich commented:
> That's a ReQL proposal that this change partially implements. r.now() becomes reevaluated for each change in .changes(). Note that as implemented it's not the timestamp at which the change happens -- it's some other timestamp that gets computed later. That's pretty weird.
The other change that this introduces is that r.now() can't be used before .changes() is called. If you want changes that refer to anything referring to a future time, or a past time, or anything over 3 weeks old, now you can't do that using 

Weirdness aside, this is very useful, it's actually a feature! But wouldn't the disallow of r.now() before changes() be very counter-intuitive(and also break production code)?

thelinuxlich commented:
If I understood correctly, this would allow something like the continuous queries of InfluxDB:

https://docs.influxdata.com/influxdb/v0.9/query_language/continuous_queries/

srh commented:
Yeah, `r.now()` before changes() would break stuff, if there's code that uses it.

After .changes(), it was always broken (it used a hard-coded value).  And then either `r.now()` or some other command name `r.write_timestamp()` could be used.  I think it should be disabled for now, because having `r.now()` change meaning depending on which side of `.changes()` is probably a bad idea.

srh commented:
Working on this.

AtnNn commented:
@srh are you still working on this?

After thinking about this for a while, I think the terminology is wrong. `is_deterministic` should return a `bool` and depend on the current environment.

If we implement that with a bitset, `term_t` could have a `nondet_set()` method and `env_t` could have a `det_set()` method, and `term->is_deterministic(env)` ⇔ `(bool)(term->nondet_set() & ~env->det_set())`

srh commented:
Oh sorry, I am still working on this and should be done today (using the technique of "actually" working on it).

I don't want to rename it just because I want to keep a clean diff for now.  But in the future I don't want to have to construct an `env_t` just to see if a query is deterministic.

srh commented:
I have fixes done in the server (in my branch: https://github.com/srh/rethinkdb/tree/sam/deterministic_now ).  Comparison functions are made to allow `r.now()`, and it is allowed before `changes()`, disallowed after `changes()`.

I guess some automatic test cases should be added too, but that's not done yet.

srh commented:
Added automatic test cases.  So that's done and up in that branch.

srh commented:
I've got a version of my branch above that has been rebased on next: https://github.com/srh/rethinkdb/tree/sam/deterministic_now_rebased

The rebasing process has some janitorial work done, related to the split-up of datum_stream.hpp.

AtnNn commented:
Thanks @srh, that `deterministic_t` looks easier to use. I'm taking a closer look at the code.

AtnNn commented:
@srh I added your changes to this PR. I also fixed a GCC 5.4 complaint about a `constant_now` name conflict between the enum  and method. Does it LGTY?

I like the idea of erroring if `r.now` is used after `changes`, but perhaps that should be a ReQL proposal first.

AtnNn commented:
I'm not sure disabling `r.now` after `.changes` is desired. There seems to be legitimate use cases for it, for example:

`.changes().map(r.row('new_val')('updatedAt').sub(r.now()))`

It also introduces confusion. With your patch `.changes().filter(_ => r.now())` fails but `.changes().filter(r.now())` works.

srh commented:
The problem with `.changes().map(r.row('new_val')('updatedAt').sub(r.now()))` is that you can come up with another example where people expect `r.now()` to refer to whatever time it is when the changefeed arrives.  And there's a workaround, `r.now().do(x => r.row('new_val')('updatedAt').sub(x)`.

AtnNn commented:
It's not innocuous (it might break existing queries) and it isn't obvious to me that is is a net improvement:

* Users might expect a different behaviour, but they can't have it. The current behaviour of `r.now()` after `changes` is documented, it is the same as everywhere else (except for write hooks) and has valid use cases.
* If users start seeing errors about `r.now()` after `changes` they might be more inclined to think that its use elsewhere isn't constant (such as after a long-running filter), because it doesn't error.
* Also, the undocumented evaluation order of ReQL combined with your change leads to inconsistent results, the like of which I would prefer to remove instead of add. For example:

Query | Result
-- | --
`.changes().filter(_ => r.now())` | Error
`.changes().filter(r.now())` | Okay
`.changes().merge(_ => ({t: r.now()}))` | Error
`.changes().merge({t: r.now()})` | Error

I would love to hear other opinions on this, which is why I suggested making it a ReQL proposal.

srh commented:
Then if you want to, make it so that `r.now()` behaves after `changes` the same as it does now.  I think the behavior's an un-thought-out artifact of implementation details of the interpreter though.

srh commented:
Note that to fix this, the commit you'll want to replace is everything in de3f25e1c5ab52a875d01e2e5d69cba2de8fa257.

srh commented:
Disallowing `r.now()` after `changes()` can get really gross when you realize that queries like this are legit:

`r.branch(r.random().gt(0.5), r.table('foo'), r.table('foo').changes()).merge({'z': r.now()})`

Disallowing `r.now()` after `changes()` would force the evaluation order, especially if we try to fix `.changes().filter(EXPR)` so that `EXPR` cannot use `r.now()` at all.  So that's another reason to just make it constant everywhere.

srh commented on src/rdb_protocol/terms/writes.cc:
```diff
@@ -313,7 +315,7 @@ class replace_term_t : public op_term_t {
         if (!nondet_ok) {
             // If this isn't true we should have caught it in the `rcheck` above.
-            rassert(f->is_deterministic() == deterministic_t::always);
+            rassert(f->is_deterministic().test(single_server::no, constant_now::yes));
```
Compilation error here.

danielmewes commented:
Am I correct that this change ended up as part of https://github.com/rethinkdb/rethinkdb/pull/6396 and we no longer need this PR?

srh commented:
Yes, that's correct.

srh commented:
Closing.
